### PR TITLE
feat(zk): ieee754 differential harness vs hardware f32/f64 oracle — PROOF M1 (sq-3x7dl.14.1) [SONNET-4.6]

### PR DESCRIPTION
**Pull request by jeswr**
> 🤖 **SPARQ agent** (sparq-rust-impl, Sonnet 4.6) — bead sq-3x7dl.14.1.

## Summary

PROOF M1: a trusted-independent differential harness that asserts bit-equality of `sparq_ieee754` Noir primitives against native Rust hardware f32/f64 + `half::f16` for f16, wired into CI with a seeded-regression self-test.

### Design record
`research/zk-correctness-and-proof-program.md` (PR #1466 / `feat/zk-correctness-proof-program`)

### Acceptance test
`scripts/run_differential_harness.sh` — executed and verified locally before this PR:
- Oracle mode: **24/24 tests passed** (4,543 assertions)
- Fault-injection self-test: `differential_oracle_f32_add` correctly **FAILS** when one expected bit is flipped — proves non-vacuity

### Files changed
- `zk/ieee754/differential/` — standalone Rust binary (NOT in root workspace; own `[workspace]` so it never drags `half` into the lean sparq-core build)
- `zk/ieee754/scripts/run_differential_harness.sh` — driver: build harness → oracle mode → nargo test → fault self-test
- `zk/ieee754/tests/differential_oracle/src/lib.nr` — committed initial version of the generated 9,201-line oracle Noir file (regenerated by CI on each run)
- `.github/workflows/zk-toolchain.yml` — new `differential-ieee754` job (zk-path filtered, includes changed-files merge-group gate, no `|| true`/`continue-on-error`)

### Oracle / TCB

| Width | Oracle | Notes |
|-------|--------|-------|
| f32/f64 | Native Rust hardware floats | IEEE 754-2008 round-to-nearest-even on any conformant CPU; the universal reference |
| f16 | `half::f16` crate | f32-widening arithmetic; correctly rounded for normal values |
| f128 | **DEFERRED** | Rust `f128` nightly-only; no stable verified soft-float binding — follow-up bead sq-3x7dl.14.2 |

**This is VERIFICATION not proof.** The trusted chain tested: Noir source → ACIR → `nargo test` witness generation. The ACIR→Barretenberg→proof lowering is NOT covered. Sample-based (32 random + corner cases per op), not exhaustive.

### Corpus
- Corner cases: ±0, ±inf, quiet/signaling NaN, min/max subnormal, min/max normal, 1.0, ties-to-even boundaries — per type
- Seeded pseudorandom: 32 pairs per arithmetic op, 20 per comparison op (seed `0x4F52_4143_4C45_F754`, xorshift64, FIXED — never time-based)
- Operations: add, sub, mul, div, sqrt (arithmetic); eq, lt, le (comparisons); for f16, f32, f64

### Seeded-regression self-test
`--inject-fault` flips the LSB of one arithmetic expected value → `nargo test` MUST fail. The shell script asserts this; if it passes, CI exits non-zero with "harness is VACUOUS".

Verified locally:
```
[differential-harness] Self-test PASSED: fault injection correctly detected.
[differential-harness] Differential harness is non-vacuous and wired to sparq_ieee754.
```

### Gates
- Standalone crate clippy `-D warnings`: ✅ (0 warnings)
- Standalone crate tests: ✅ (14/14 pass)
- Rustdoc `--all-features -D warnings`: ✅
- `scripts/check-readme-template.py --enforce`: ✅ (0 deviations, 48 READMEs)
- Main workspace clippy/tests: ✅ (standalone crate not in workspace — no interference)
- YAML lint: ✅
- End-to-end nargo test (oracle): ✅ 24/24
- End-to-end nargo test (fault): ✅ correctly fails

**Auto-merge: OFF** — awaiting independent verify per bead contract.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: zk/ieee754/differential/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,1066 @@
...
+                writeln!(out, "    // corner:{}", idx).unwrap();
+                writeln!(out, "{}", f16_cmp_line(a, b, expected, method)).unwrap();
+                idx += 1;
+            }
+        }
+
+        for i in 0..RANDOM_CMP_PER_OP {
+            let a = rng.next_u16() & 0x7BFF;
+            let b = rng.next_u16() & 0x7BFF;
+            let expected = oracle_fn(a, b);
```
[the remark this answers is not in the document]
Comment (jeswr) [reply zk/ieee754/differential/src/main.rs:816]: > 🤖 **SPARQ agent** — fixed in commit a4c005f7.

Changed `rng.next_u16() & 0x7BFF` to `& 0xFBFF` in `gen_f16_cmps`. The old mask `0x7BFF` cleared both the sign bit (bit 15) and the exponent LSB (bit 10), making all random cmp pairs positive-only. The new mask `0xFBFF` clears only bit 10 (prevents inf/NaN) while preserving bit 15, mirroring the f32 mask `0xFF7FFFFF` (clears f32 exponent LSB, bit 23) and f64 mask `0xFFEFFFFFFFFFFFFF` (clears f64 exponent LSB, bit 52).

Added a direct unit test `test_f16_cmp_mask_preserves_sign` that asserts the mask property and flags the old mask as incorrect.

File: zk/ieee754/scripts/run_differential_harness.sh [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    exit 1
+fi
+if ! command -v nargo &> /dev/null; then
+    echo "ERROR: nargo not found -- see .github/workflows/zk-toolchain.yml for install" >&2
+    exit 1
+fi
+
+echo "[differential-harness] Building Rust harness..."
+cargo build --manifest-path "${HARNESS_DIR}/Cargo.toml" --release 2>&1
+HARNESS_BIN="${HARNESS_DIR}/target/release/sparq-ieee754-differential"
```
[the remark this answers is not in the document]
Comment (jeswr) [reply zk/ieee754/scripts/run_differential_harness.sh:54]: > 🤖 **SPARQ agent** — fixed in commit a4c005f7.

`--locked` added to the `cargo build` step in `run_differential_harness.sh`. Also added a standalone CI step "Drift guard" that runs `cargo build --locked` independently and diffs the generator output against the committed `tests/differential_oracle/src/lib.nr`, failing on any drift without `|| true`.
